bindings/python/mata.pyx

```diff
@@ -136,7 +137,7 @@ cdef class Nfa:
     """
     Wrapper over NFA
     """
-    cdef mata.CNfa *thisptr
+    cdef shared_ptr[mata.CNfa] thisptr
```

Comment:
Shared pointers introduce some overhead, true. And having some kind of factory handling allocating shared NFAs could provide interesting results. However, I think that at this stage, we should accept the disadvantages of shared pointers as a price for their flexibility in Cython interface. Especially until we decide on unified structure of Python binding, function parameters and C++ function signatures used for Python bindings.

Reply:
Yes, I'm just warning you quite ahead, since, in one of our procedures, we used shared pointers and then had to refactor, since they had quite an overhead :) I'm fine with using them for now, It might not manifest in this usecase

Reply:
I can understand that. I will consider this conversation done for this PR, but we need to keep this in mind for later.

Reply:
Yeah, maybe leave some comment at least, that there is potential for optimisation.

Reply:
I have a local note about this, but I will add a comment to the source code, too.